src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs

```diff
@@ -23,18 +25,40 @@ internal abstract class AnalyzerAssemblyLoader : IAnalyzerAssemblyLoader
     {
         private readonly object _guard = new();
 
-        // lock _guard to read/write
-        private readonly Dictionary<string, Assembly> _loadedAssembliesByPath = new();
+        /// <summary>
+        /// Set of analyzer dependencies original full paths to the data calculated for that path
+        /// </summary>
+        /// <remarks>
+        /// Access must be guaraded by <see cref="_guard"/>
+        /// </remarks>
+        private readonly Dictionary<string, (AssemblyName? AssemblyName, string RealAssemblyPath)?> _analyzerAssemblyInfoMap = new();
 
-        // maps file name to a full path (lock _guard to read/write):
+        /// <summary>
+        /// Maps analyzer dependency simple names to the set of original full paths it was loaded from. This _only_ 
+        /// tracks the paths provided to the analyzer as it's a place to look for indirect loads. 
+        /// </summary>
+        /// <remarks>
+        /// Access must be guaraded by <see cref="_guard"/>
+        /// </remarks>
         private readonly Dictionary<string, ImmutableHashSet<string>> _knownAssemblyPathsBySimpleName = new(StringComparer.OrdinalIgnoreCase);
 
         /// <summary>
-        /// Implemented by derived types to actually perform the load for an assembly that doesn't have a cached result.
+        /// The implementation needs to load an <see cref="Assembly"/> with the specified <see cref="AssemblyName"/>. The
+        /// <paramref name="assemblyOriginalPath"/> parameter is the original path. It may be different than
+        /// <see cref="AssemblyName.CodeBase"/>
         /// </summary>
-        protected abstract Assembly LoadFromPathUncheckedImpl(string fullPath);
+        /// <remarks>
+        /// The implementation should return null on error, not throw an exception
+        /// </remarks>
+        protected abstract Assembly Load(AssemblyName assemblyName, string assemblyOriginalPath);
```

Comment:
I don't see any nullable directives in the file, isn't nullable enabled at this location?

Reply:
I need to update that comment. This code originally indicated `null` was valid. After I dug into the API call tree though I figured out `null` wasn't a valid return. It either loads or throws.